warpforge-visualize/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+	fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+		writeln!(f, "digraph {{")?;
+		fmt_entries(&self.0, f, IDENT_SPACES)?;
+		writeln!(f, "}}")?;
+		Ok(())
+	}
+}
+
+fn fmt_entries(entries: &[Statement], f: &mut Formatter<'_>, ident: usize) -> std::fmt::Result {
+	let prefix = " ".repeat(ident);
+	for statement in entries {
+		match statement {
+			Statement::Node(node) => {
+				write!(f, "{prefix}\"{}\"", node.name)?;
+				for attribute in &node.attributes {
+					write!(f, "{attribute}")?;
+				}
+				writeln!(f, ";")?;
+			}
+			Statement::Edge(edge) => {
+				write!(f, "{prefix}\"{}\" -> \"{}\"", edge.from, edge.to)?;
+				for attribute in &edge.attributes {
+					write!(f, "{attribute}")?;
+				}
+				writeln!(f, ";")?;
+			}
+			Statement::SubGraph(name, graph) => {
+				writeln!(f, "{prefix}subgraph \"cluster_{name}\" {{")?;
+				fmt_entries(&graph.0, f, ident + IDENT_SPACES)?;
+				writeln!(f, "{prefix}}}")?;
+			}
+		}
+	}
+	Ok(())
+}
+
+pub(crate) trait Graphable {
+	fn graph(&self) -> Graph;
+}
+
+impl Graphable for Plot {
+	fn graph(&self) -> Graph {
+		let mut graph = Graph::default();
+
+		let mut inputs_graph = Graph::default();
+		for (label, _) in &self.inputs {
+			inputs_graph
+				.add_node(Node::new(format!("pipe::{}", label.0)).with_label(label.0.to_string()));
+		}
+		graph.add_subgraph("inputs".to_string(), inputs_graph);
+
+		for (label, step) in &self.steps {
+			let mut subgraph = Graph::default();
+			let step_label = label.0.to_string();
+			let Step::Protoformula(formula) = step else { unimplemented!() }; // TODO
+			for (target, input) in &formula.inputs {
+				if let PlotInput::Pipe(_) = input {
+					let input_target_id = format!("{step_label}{}", target.0);
+					subgraph
+						.add_node(Node::new(input_target_id.clone()).with_label(target.0.clone()));
+					subgraph.add_edge(
+						Edge::new(input.to_string(), input_target_id.clone())
+							.with_arrow(ArrowType::Empty),
+					);
+					subgraph.add_edge(Edge::new(input_target_id, step_label.clone()));
+				}
+			}
+			subgraph.add_node(Node::new(step_label.clone()).with_shape(NodeShape::Box));
+			graph.add_subgraph(step_label, subgraph);
+		}
+
+		for i in 1..self.steps.len() {
+			graph.add_edge(Edge::new(
+				self.steps.get_index(i - 1).unwrap().0 .0.to_string(),
+				self.steps.get_index(i).unwrap().0 .0.to_string(),
+			));
+		}
+
+		graph
+	}
+}
+
+pub(crate) fn fetch_warpsys_plot(base_url: &Url, package: &str) -> Option<Plot> {
+	let mut url = base_url.clone();
+	url.path_segments_mut()
+		.unwrap()
+		.push(package)
+		.push(PLOT_FILENAME);
+	eprintln!("{url}");
+	let plot_file = reqwest::blocking::get(url.clone()).unwrap();
+	if plot_file.status() != StatusCode::OK {
+		return None;
+	}
+	let parse_result = serde_json::from_reader(plot_file);
+	if parse_result.is_err() {
+		eprintln!("Failed to parse '{url}'");
+	}
+	let PlotCapsule::V1(plot) = parse_result.unwrap();
+	Some(plot)
+}
+
+pub(crate) fn fetch_warpsys_plot_recursively(
```

Comment:
I can see this being handy :D

I think you could pretty much `s/warpsys_//g` out of the function names across the whole area, though.  This is all data layout that's general to all catalogs, and warpsys is just an instance of a catalog.

Reply:
I think it would be great to have a catalog reading module that contains functions to do this (and other kinds of lookups) on either a filesystem-backed implementation or a url-backed implementation.

Not sure how far you want to drive towards that right now in the same PR, though.  I'd be fine seeing this code land now and move around later as we figure it out more.

Reply:
I think it could be nice to have some functions/api like that in the `warpforge-api` project. But I intentionally only did prototypy code here. For example I would do parallel download with something like tokio if I had to port this to production-like code.